services/QuillLMS/app/models/activity_session.rb

```diff
@@ -93,6 +94,7 @@ class StudentNotAssignedActivityError < StandardError; end
   before_save :set_completed_at, :set_activity_id
 
   after_save :determine_if_final_score, :update_milestones, :increment_counts
+  after_save :handle_partner_session_callbacks, if: -> { marking_as_complete? }
```

Comment:
Take or leave, you might consider a `if: -> { marking_as_complete? && has_partner_session? }`